src/libsyntax/parse/parser.rs

```diff
@@ -3337,6 +3339,56 @@ impl<'a> Parser<'a> {
         Ok(self.mk_expr(lo.to(hi), ExprKind::IfLet(pats, expr, thn, els), attrs))
     }
 
+    /// With eRFC 2497, we need to check whether an expression is ambigious and warn or error
+    /// depending on the edition, this function handles that.
+    fn while_if_let_ambiguity(&self, expr: &P<Expr>) {
+        if let Some((span, op_kind)) = self.while_if_let_expr_ambiguity(&expr) {
+            let message = format!("ambigious use of `{}`", op_kind.to_string());
+            let mut err = if self.span.edition() >= Edition::Edition2018 {
+                self.diagnostic().struct_span_err(span, &message)
```

Comment:
As we noted on the issue, let's make this an unconditional error for now. Otherwise, we would want to rewrite this to use a lint anyhow. In that case, I think we would *probably* want to move this code over into the [ast validator](https://github.com/rust-lang/rust/blob/master/src/librustc_passes/ast_validation.rs), since then we would have node-ids we could use.

Reply:
I've moved this to ast validation and made it a unconditional error.